### PR TITLE
Cppclass overloadable methods

### PR DESCRIPTION
**Pull request by samaingw**
This is a subset of PR #2951, containing the code concerning overloading of defined methods in cppclass scope.

No problem to rework the test if it doesn't look good.

For the records, you'll find below the original #2951 description of theses changes:

-------------------

## 2 - Overloaded methods
The goal here is to be able to define (and not just declare) methods in CppClass that can be overloaded.
I will now take some time to explain why the current implementation cannot do this, and why I think the approach proposed in my commits is a not-so-bad one.

TL;DR: The current code doesn't allow clean method inheritance.

In the current implementation of `CClassScope`, when we declare a cfunction, we check for the presence of an entry with the same name, and if this is the case, we shout if the function signature is different or if the entry is already defined.
So we cannot have declaration and/or definition of different function types with the same name.
You can declare then define a function, as it's the same function signature, so everything is fine.

In the current implementation of `CppClassScope`, when we declare a cfunction, we check for the presence of an entry with the same name and if we try to define, not just declare, the new entry.
If this is the case, we update the type if the signatures (function types) are compatible, else we shout.
In any other cases (no previous entry or new entry is just declared, not defined), override the entry.
This checks are performed in `CppClassScope.declare_var`. This also explains why the cpp methods are in the `var_entries` array.
Polymorphism is usable because in `CppClassScope.declare_cfunction` (which calls the above-mentioned `declare_var`), we link the previously declared entries to the new one if we're just declaring the entry, not defining it.
Moreover, in this `declare_cfunction`, we never touch `cfunc_entries`. This explains why this array is always empty in a `CppClassScope`.
For a classic (non-special) method, the flow is:
CppClassScope.declare_cfunction (will link the alternatives to have polymorphism)
----> calls CppClassScope.declare_var (make the checks, if we create a new entry, add it to var_entries)
-----------> calls Scope.declare (blindly create and override the entry in the entries dict)

Wrapping C++ class is possible, because we fall into the "override the entry" clause (we never define an entry).

A real issue is method inheritance.
When inheriting methods with the same name, only one is kept because of the continuous override in `Scope.declare`.
The kept method will be, in my current understanding of the code, the last (in the base) upper (not deep in the inheritance chain: the "parenthesied" bases, not base of base) right-most (in the parenthesis) base method with this name:
```
cdef cppclass Base1:
    int method(int a)

cdef cppclass Base2:
    int method(double a)

cdef cppclass HiddenBase:
    int method(double b, double b)

cdef cppclass Base3(HiddenBase):
    int method(double b, int a)
    int method(int a, double b) # <=== This one will be the only method kept in Derived by inheritance

cdef cppclass Derived(Base1, Base2, Base3):
    pass
    # The only method Derived is aware of is Base3.method(int, double), the other ones were (sadly) lost in the process
```
So we may have lost a lot of alternatives.
This is exactly what is described in issue #1357.

The proposed approach deals with the alternative management directly in `Scope.declare`.
The behaviour is quite simple: we scan the previous alternatives for a compatible function signature.
If there is a match and the previous method is marked as inherited (or the default init), we override it.
If there is a match and it's not an inherited method, shout.
If there is no match, add the new entry to the alternatives.

This involves six changes:
a) Change `Symtab.declare` logic for cppclass methods: the `overloaded_entry` array is properly managed in this code (the switch from is_cpp to is_cpp_class_scope is intentional: it restricts this overloading to cppclass, otherwise it is available to everything when cython produces c++ code)
b) Change the `declare_var` and `declare_cfunction` in CppClassScope to use the previous changes
c) Add the `overloaded_alternative` array to ubcm_entry created with a Type.method() statement
d) Amend a test case (cpp_no_constructor) because the error message is not the same:
Below is a long text explaining why this is not a stupid idea to modify this test.
Beware, there is e) and f) change sets under it.

TL;DR: Method alternatives order in the entries corresponds to the declaration order. Before the patch, the alternatives order corresponded to reversed declaration order.

`PyrexTypes.best_match` shouts when it does not find a suitable alternative.
When doing so, it uses the "principal alternative" as a reference.
Principal alternative here refers to the entry which is in `entries[name]` in the type scope.
The other ones are held in the `overloaded_alternatives` array of this principal alternative.

Before the a) change set, alternatives in a `CppClassScope` are always overridden, and it is the duty
of `CppClassScope.declare_cfunction` to fill the `overloaded_alternative` array of the newly created alternative.
So the principal alternative is always changing, and the last alternative declared will effectively be the principal one.
Consequently, when `PyrexTypes.best_match` shouts, it uses the last declared alternative as a reference (for argument number and such).

After the a and b change set, `Scope.declare` takes care of the alternatives, and overrides an alternative in a `CppClassScope` only if it is an inherited one with the same signature as the new one. As the signature is the same, each info used by `PyrexTypes.best_match` when shouting doesn't change between old overridden alternative and new one.
Consequently, a signature represented by an alternative never change position: if it is carried by the principal alternative, it will always be carried by the principal one, therefore `PyrexTypes.best_match`, when taking the principal alternative as a reference, is effectively taking the first defined alternative as reference.

As `cpp_no_constructor` is an error test about failing to find the correct alternative, it expects `PyrexTypes.best_match` to shout.
The expected string refers to the last (2nd / 2 defined alternatives) alternative, whereas with a & b changes, it should refer to the first alternative, because that's what is expected.

This commit does this.

e) Amend a test case (cpp_classes_def) because one warning must stand:
It is, again, a long text explaining why I'm not mad.

TL;DR: The test module lacked a _WARNINGS variable to catch the RegularPolygon.area warning since the beginning. It's just that it wasn't noticeable before.

In `cpp_classes_def` (in the tests/run directory), we have some defined method (area) in a class (RegularPolygon) which has no exception indication, but is performing a division.
Consequently, a divide-by-zero check is performed, and we should, theorically, raise an exception if the check is true.
As there is no except indication on this method, we have a compile-time warning : Unraisable exception in function 'RegularPolygon.area'.
This is not a bug, it's a feature.

So we should have a warning from cython when compiling this test module. A warning that should be reported when running the test suite.
With the a) & b) changes in Symtab, this is what's happening.
This wasn't before, and the reason is quite tricky.

`RegularPolygon` has a base class, `Shape`, which is declared as 'extern'. Shape has the `area` method.
So, RegularPolygon inherits area from Shape before overloading it.
`Shape.area` and `RegularPolygon.area` have exactly the same signature.
In `CppClassScope.declare_var`, we are in the following situation: previous entry and defining.
This leads to:
*We update the type if the signatures (function types) are compatible, else we shout*
As types are compatible, we just modify the entry type, keeping the inherited entry for all other attributes.
The side effect is that we didn't change `entry.pos`. An inherited entry has a pos equal to None, because the corresponding attribute is declared nowhere in the subclass body.

Consequently, when making the warning, the entry has no pos therefore it is discarded by the test suite when trying to match actual warnings from expected ones.

This commit adds the warning from RegularPolygon.area to the expected warnings. It also removes werror to the tags, since this warning is not a runtime blocker.

f) Provide test for the feature (which is failing on the current master as it is related to issue #1357)

## 3 - Cppclass generation fixes
a) Allow defined (not wrapped) cppclass to have multiple constructors
This fixes cppclass generation. With the above code everything is there to handle overloaded constructors.
This is about generating the correct C++ code when we're not wrapping an external cppclass, but defining it in cython.

### CODE REVIEW

**Comment by samaingw** [comment]
Sorry for the PR spam, but I've got some reproductibility issues on my own machine, so there is still some errors I have troubles checking locally.

**Comment by samaingw** [comment]
@scoder : Do you want anything more from me on this matter ?

File: tests/run/cpp_classes_def.pyx [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # mode: run
-# tag: cpp, werror, cpp11
```
Comment (scoder) [inline tests/run/cpp_classes_def.pyx]: In order to read the warnings below, this should include the tag `warnings`.

File: Cython/Compiler/Symtab.py [hunk 2]
```diff
@@ -2430,16 +2442,19 @@ def declare_var(self, name, type, pos,
...
-            elif type.is_cfunction and type.compatible_signature_with(entry.type):
-                entry.type = type
             else:
                 error(pos, "Function signature does not match previous declaration")
         else:
             entry = self.declare(name, cname, type, pos, visibility)
+            if type.is_cfunction and not defining:
+                entry.is_inherited = 1
         entry.is_variable = 1
+        entry.is_cfunction = type.is_cfunction
```
Comment (scoder) [inline Cython/Compiler/Symtab.py]: I'd only assign this `if type.is_cfunction`. There can be a lot of Entries, so saving instance dict space is a good thing.

File: Cython/Compiler/Symtab.py [hunk 3]
```diff
@@ -470,16 +470,21 @@ def declare(self, name, cname, type, pos, visibility, shadow = 0, is_type = 0, c
...
+        old_index = -1
         if name and name in entries and not shadow:
             old_entry = entries[name]
 
             # Reject redeclared C++ functions only if they have the same type signature.
             cpp_override_allowed = False
             if type.is_cfunction and old_entry.type.is_cfunction and self.is_cpp():
-                for alt_entry in old_entry.all_alternatives():
-                    if type == alt_entry.type:
+                for index, alt_entry in enumerate(old_entry.all_alternatives()):
```
Comment (scoder) [inline Cython/Compiler/Symtab.py]: This makes an assumption about the order returned by `.all_alternatives()`. The code here shouldn't rely on that.

File: Cython/Compiler/Symtab.py [hunk 4]
```diff
@@ -2509,6 +2525,9 @@ def declare_inherited_cpp_attributes(self, base_class):
                     base_entry.type, None, 'extern')
                 entry.is_variable = 1
                 entry.is_inherited = 1
+                entry.is_cfunction = base_entry.is_cfunction
+                if entry.is_cfunction:
+                    entry.func_cname = base_entry.func_cname
```
Comment (scoder) [inline Cython/Compiler/Symtab.py]: I suggest making the whole thing depend on `if base_entry.is_cfunction`.

**Comment by samaingw** [comment]
Thanks for your feedback, I'll process it ASAP.

File: Cython/Compiler/Symtab.py (hunk 4 above)
Comment (samaingw) [reply Cython/Compiler/Symtab.py]: Done, thanks for the suggestion.

File: Cython/Compiler/Symtab.py (hunk 2 above)
Comment (samaingw) [reply Cython/Compiler/Symtab.py]: Thank you for this remark, I wasn't thinking about memory consumption when I made this. Fixed.

File: tests/run/cpp_method_overloading.pyx [hunk 5]
```diff
@@ -0,0 +1,77 @@
+# mode: run
+# tag: cpp
+
+cdef extern from "cpp_method_overloading_support.h":
```
[the remark this answers is not in the document]
Comment (samaingw) [reply tests/run/cpp_method_overloading.pyx]: Thank you for the tip ! It's corrected now.

File: Cython/Compiler/Symtab.py (hunk 3 above)
Comment (samaingw) [reply Cython/Compiler/Symtab.py:480]: I understand the point. The goal here is to mark a given inherited alternative as not inherited anymore. I see two possible ways:

1. Create a new entry as the rest of the cases and override the correct alternative with this new entry
2. Update the current entry without creating a new one

I've chosen option 1, and as I need to override the correct alternative in the overloaded_alternatives array, the index pops out.
I'm perfectly OK with option 2 if you think it fits better.
That would consists in updating flags, cname and type of the correct alternative, and avoid creating a useless new entry.

**Comment by navytux** [comment]
> Thanks for your feedback, I'll process it ASAP.

@samaingw, are you done with adjusting changes according to review feedback?

For the reference it usually makes sense to post something like "@scoder, please take another look" as there are likely no notifications on github when pull request is updated with new code, and so reviewer might be thinking that it is still patch author turn to make or complete.

**Comment by samaingw** [comment]
@scoder : After a huge hiatus between two partial updates, it's good for a second review.

**Comment by samaingw** [comment]
@da-woods : I'm puzzled by the fact only one specific platform/python version/backend language  fails (macos-10.15, py 3.6, c, 10.14). Is this a known issue elsewhere ?
If I understand the log correctly, this is `test_asyncgen.AsyncGenAsyncioTest.test_async_gen_asyncio_gc_aclose_0` which is failing. This is apparently a specific test for python3.6, and tests done with ubuntu + py36 are all fine.
So, is this an async bug related to macos + py36 ?
I wonder if this was encountered before.

**Comment by da-woods** [comment]
Closing and reopening to rerun to the CI.

It looks like something that's just glitched rather than something related to the PR. But we shall see

**Comment by samaingw** [comment]
Looks like this failed test was just a glitch. Would you like to have comments explaining the different code chunks, like what I've done for #3221 ?
These comments won't land in a complete form soon (if I remember correctly, there is quite a lot to comment ;)), but I can definitely spare some time to make that happen.

File: tests/run/cpp_method_overloading.pyx [hunk 6]
```diff
@@ -0,0 +1,88 @@
+# mode: run
+# tag: cpp
+
+cdef extern from *:
```
Comment (da-woods) [inline tests/run/cpp_method_overloading.pyx:4]: This file could use a bit of clarity about what you're actually aiming to show with all of the tests. `Derived` doesn't really seem to do anything. Is `DefinedDerived` only to test that you can override a constructor? If so would some print statements in the constructor help to prove that it was actually called?

**Review by da-woods — COMMENTED** [review]
> Would you like to have comments explaining the different code chunks, 

I didn't find it hugely unclear looking through it. Add anything you think is useful, but don't go to excess.

I've added some other comments though.

File: Cython/Compiler/ModuleNode.py [hunk 7]
```diff
@@ -1057,30 +1057,17 @@ def generate_cpp_class_definition(self, entry, code):
...
-                    if constructor.type.optional_arg_count:
-                        arg_decls.append(constructor.type.op_arg_struct.declaration_code(Naming.optional_args_cname))
-                        arg_names.append(Naming.optional_args_cname)
-                    if not arg_decls:
-                        arg_decls = ["void"]
-                else:
-                    arg_decls = ["void"]
-                    arg_names = []
+
+            def generate_cpp_constructor_code(arg_decls, arg_names, is_implementing, py_attrs, constructor):
```
[the remark this answers is not in the document]
Comment (samaingw) [reply Cython/Compiler/ModuleNode.py:1070]: Yep, you're right. Given that `generate_cpp_constructor_code` is a helper function used only in `generate_cpp_class_definition`, it would be sensible to put it at the same scope, right ? That would be class scope then.

File: Cython/Compiler/Symtab.py [hunk 8]
```diff
@@ -493,17 +493,28 @@ def declare(self, name, cname, type, pos, visibility, shadow = 0, is_type = 0, c
...
         entries = self.entries
+        entry = None
+        cpp_already_overridden = False
         if name and name in entries and not shadow and not self.is_builtin_scope:
             old_entry = entries[name]
 
             # Reject redeclared C++ functions only if they have the same type signature.
             cpp_override_allowed = False
             if type.is_cfunction and old_entry.type.is_cfunction and self.is_cpp():
                 for alt_entry in old_entry.all_alternatives():
```
[the remark this answers is not in the document]
Comment (samaingw) [reply Cython/Compiler/Symtab.py:504]: Indeed, that's not super-clear at first sight. A comment proposition will go into the next batch.

File: Cython/Compiler/Symtab.py [hunk 9]
```diff
@@ -493,17 +493,28 @@ def declare(self, name, cname, type, pos, visibility, shadow = 0, is_type = 0, c
...
-                    if type == alt_entry.type:
+                    if type.compatible_signature_with(alt_entry.type):
                         if name == '<init>' and not type.args:
                             # Cython pre-declares the no-args constructor - allow later user definitions.
                             cpp_override_allowed = True
+                        elif alt_entry.is_inherited:
+                            cpp_override_allowed = True
+                        if cpp_override_allowed:
+                            alt_entry.type = type
```
[the remark this answers is not in the document]
Comment (samaingw) [reply Cython/Compiler/Symtab.py:512]: By "this line" I understand `alt_entry.type = type`. We can override methods which have a compatible signature in C++ (hence the change in line 505). But compatible doesn't mean exactly the same type.
So here, for the `alt_entry` which matches the redefinition, we update it's attributes to match the redefinition we are declaring, so that the updated entry looks exactly like if we didn't inherit anything.

[the remark this answers is not in the document]
Comment (samaingw) [reply Cython/Compiler/Symtab.py:512]: A comment for that will end up in the next batch then ;)

File: tests/run/cpp_classes_def.pyx [hunk 10]
```diff
@@ -269,3 +269,7 @@ def test_CppClassWithDocstring():
     cdef CppClassWithDocstring *c = new CppClassWithDocstring()
     del c
     print "OK"
+
+_WARNINGS="""
+24:4: Unraisable exception in function 'RegularPolygon.area'.
```
[the remark this answers is not in the document]
Comment (samaingw) [reply tests/run/cpp_classes_def.pyx:274]: This comes from `Symtab.Scope.declare()`, more precisely line 515 `alt_entry.pos = pos`.
`RegularPolygon` inherits from `Shape`. So `RegularPolygon.area()` is an inherited method which is redefined.

Before this PR changes, the entry corresponding to `RegularPolygon.area()` had a `pos` attribute equal to `None`,
because inherited methods are declared this way (inherited methods are inside the `inherited_var_entries` attribute).

With this PR, the inherited method is still declared with `pos = None`, but we update `pos` as soon as we have a redefinition of `RegularPolygon.area()`.

In other words, we changed from:
```
Unraisable exception in function 'RegularPolygon.area'.
```
to:
```
24:4: Unraisable exception in function 'RegularPolygon.area'.
```

Now, why does this change the test suite results ?
Because warnings without positions are discarded by the test suite.
I guess the exact place where it's done is through the necessary position groups in the
`match_error` regex of `ErrorWriter` class in `runtests.py`. (line 607)
Warnings with no position aren't matched by this regex (if I understand it correctly), so they are silently discarded.

File: Cython/Compiler/Symtab.py [hunk 11]
```diff
@@ -2738,6 +2757,13 @@ def specialize(self, values, type_entry):
 
         return scope
 
+    def lookup_here(self, name):
+        if name == "__init__":
+            name = "<init>"
+        elif name == "__dealloc__":
+            name = "<del>"
```
[the remark this answers is not in the document]
Comment (samaingw) [reply Cython/Compiler/Symtab.py:2767]: I agree. I just followed what's done in `CppClassScope.declare_cfunction`. It's easy to change, but will break code from Cython users relying on the `__init__`/`__dealloc__` name pair.
As C++ support in Cython is said to be experimental for defining C++ code, a possible dirty workaround is:
- allow a more sensible name pair (`__init__`/`__del__` ?) and make the correct wiring for that one,
- allow `__init__`/`__dealloc__` pair for wrapping C++ only (if I remember my Cython correctly, we can isolate these cases with the defining/visibility flags).

**Comment by samaingw** [comment]
@da-woods: I need your lights on the last problem when running unit test.

The error I've got triggers with this block of code in my test codes:
```cython
cdef extern from *:
    """
    class Base {
    public:
        int foo(int a) {
            return a + 1;
        }
    };
    """
    cdef cppclass Base:
        __init__()
        int foo(int)

cdef cppclass Derived(Base):
    int foo(int a):
        return a * 2;
```

What happens is:
- `Base.foo()` being wrapping an external C++ code, no exception check is performed: its signature is `int foo(int) noexcept`,
- `Derived` inherits that method, with the same signature,
- the new definition of `Derived.foo()` is declared with `exception_check=True`, following the mapping in `Parsing:p_exception_value_clause`,
- after type analysis, the signature of the new definition is `int foo(int) except? -1`,
- the inherited method is an `int foo(int) noexcept`, whereas the new method is an `int foo(int) except? -1` => these are incompatible signature types => both entries live in `Derived` scope (ie the inherited entry is *not* replaced),
- when calling `Derived.foo(int)`, cython rightfully complains because it cannot choose between both signatures.

Quite logically, if I change the code to:
```cython
cdef extern from *:
    """
    class Base {
    public:
        int foo(int a) {
            return a + 1;
        }
    };
    """
    cdef cppclass Base:
        __init__()
        int foo(int)

cdef cppclass Derived(Base):
    int foo(int a) noexcept:
        return a * 2;
```
then everything works like a charm.

My question is: is this a bug or a feature ?
In other words, when redefining methods for a C++ class deriving from an external C++ base class, should we:
1. assume the exception handling is the same as the base class method being overloaded, or
2. shout as the current code does ?

**Comment by da-woods** [comment]
> shout as the current code does ?

I think it makes sense to shout. In this case `Base::foo` isn't virtual, but generally it might be and Cython doesn't know. That means any time it calls `Base::foo` it could be calling `Derived::foo` and thus it would behave wrongly with respect to exceptions. If possible, it'd be good to make the error message clear - the implicit exception specification is not always obvious to see.

I'm going to take a proper look at this PR when I can - probably next weekend.

**Comment by samaingw** [comment]
Failures happen on windows-2019 py3.12 (both c and cpp), with test `test_cycache_uses_cache`, which doesn't seem really linked to this PR => closing and reopening to see if this was just a glitch.

**Comment by samaingw** [comment]
Failures happen on windows-2019 py3.12 (both c and cpp), with test `test_cycache_uses_cache`, which doesn't seem really linked to this PR => closing and reopening to see if this was just a glitch.

**Comment by samaingw** [comment]
Failures happen on windows-2019 py3.12 (both c and cpp), with test `test_cycache_uses_cache`, which doesn't seem really linked to this PR => closing and reopening to see if this was just a glitch.

**Comment by samaingw** [comment]
Failures happen on windows-2019 py3.12 (both c and cpp), with test `test_cycache_uses_cache`, which doesn't seem really linked to this PR => closing and reopening to see if this was just a glitch.

**Comment by samaingw** [comment]
Failures happen on windows-2019 py3.12 (both c and cpp), with test `test_cycache_uses_cache`, which doesn't seem really linked to this PR => closing and reopening to see if this was just a glitch.

**Comment by samaingw** [comment]
@da-woods: The PR is OK for review on my end.
I've noted your remark on the not-so-obvious error message:
> If possible, it'd be good to make the error message clear - the implicit exception specification is not always obvious to see.

From my end, I will allocate very few time for this PR for, probably, the next 2 weeks or so.
I already have an idea for giving a more explicit error message, however my current idea is a bit clunky.
I'll search for a way to do that in a cleaner way. That'll end up either in a new PR if you decided to merge this one,
or this PR if I've got some homework after your review ;)

File: Cython/Compiler/ModuleNode.py [hunk 12]
```diff
@@ -1033,6 +1033,21 @@ def generate_struct_union_definition(self, entry, code):
...
+        if is_implementing:
+            code.putln("%s(%s) {" % (type.cname, ", ".join(arg_decls)))
+            if py_attrs:
+                code.put_ensure_gil()
+                for attr in py_attrs:
+                    code.put_init_var_to_py_none(attr, nanny=False)
+            if constructor:
+                code.putln("%s(%s);" % (constructor.cname, ", ".join(arg_names)))
+            if py_attrs:
+                code.put_release_ensured_gil()
```
Comment (da-woods) [inline Cython/Compiler/ModuleNode.py:1046]:
```suggestion
            needs_gil = py_attrs or (constructor and not constructor.type.nogil)
            if needs_gil:
                code.put_ensure_gil()
            if py_attrs:
                for attr in py_attrs:
                    code.put_init_var_to_py_none(attr, nanny=False)
            if constructor:
                code.putln("%s(%s);" % (constructor.cname, ", ".join(arg_names)))
            if needs_gil:
                code.put_release_ensured_gil()
```

I realise this is just code you've moved, but probably worth fixing while we're touching it - I think it's worth getting the GIL if the constructor needs it too.

File: Cython/Compiler/Symtab.py [hunk 13]
```diff
@@ -493,17 +493,35 @@ def declare(self, name, cname, type, pos, visibility, shadow = 0, is_type = 0, c
             warning(pos, "'%s' is a reserved name in C." % cname, -1)
 
         entries = self.entries
+        entry = None
+        cpp_already_overridden = False
         if name and name in entries and not shadow and not self.is_builtin_scope:
             old_entry = entries[name]
```
Comment (da-woods) [inline Cython/Compiler/Symtab.py:499]: I've proposed an update for this section in https://github.com/cython/cython/commit/a440bd688b45a40919d76be50c167d253ff248b6.

I've tried very hard not to change your underlying logic. However, it's making "declare" quite complicated.

Instead I'm arguing that most of this can be moved into `CppClassScope` leaving the base class declare to be relatively simple. Let me know if that seems reasonable - feel free to modify it.

**Review by da-woods — COMMENTED** [review]
I think I basically believe this works, and would be happy to merge something similar.

I've proposed a change (I've linked to the commit in one of the comments above) to move most of the logic out of "declare").

I think that lowers the impact of this change a bit because at least we know that any issues will be limited to C++ classes.

**Comment by da-woods** [comment]
Sorry this hasn't been dealt with hugely quickly.  Will try to get this merged fairly promptly after you've looked at my remaining comments.

**Comment by da-woods** [comment]
Right - I've pushed my remaining review suggestions to this branch. I'll leave it about a week for comments, then merge it.

**Comment by samaingw** [comment]
Sorry for the huge delay from my part !
I'm happy with moving the re-declaration logic in a separate method, this is all good for me.

Thanks for the great work !

**Comment by samaingw** [comment]
@da-woods : you're good to go with this branch?

**Comment by da-woods** [comment]
Thanks - sorry this took a while to deal with.

File: Cython/Compiler/Symtab.py [hunk 14]
```diff
@@ -501,6 +501,66 @@ def new_c_type_context(self, in_c_type_context=None):
...
+                    if cpp_override_allowed:
+                        # A compatible signature doesn't mean the exact same signature,
+                        # so we're taking the new signature for the entry.
+                        alt_entry.type = type
+                        alt_entry.is_inherited = False
+                        # Updating the entry attributes which can be modified in the method redefinition.
+                        alt_entry.cname = cname
+                        alt_entry.pos = pos
+                        entry = alt_entry
+                    break
```
Comment (scoder) [inline Cython/Compiler/Symtab.py:541]: I'm not sure if the order of `all_alternatives()` is always deterministic. I would guess that this may lead to fluctuation in the generated C code, e.g. when using overrides with different C integer types. We should definitely prefer exact matches over close signature matches. There is already code that tries to determine the best matching C++ signature. I think we should reuse that here.

File: Cython/Compiler/Symtab.py [hunk 15]
```diff
@@ -501,6 +501,66 @@ def new_c_type_context(self, in_c_type_context=None):
         yield
         self.in_c_type_context = old_c_type_context
 
+    def handle_already_declared_name(self, name, cname, type, pos, visibility):
```
Comment (scoder) [inline Cython/Compiler/Symtab.py:504]: Maybe `find_declared_entry()` would be a better name.

File: tests/errors/cpp_no_constructor.pyx [hunk 16]
```diff
@@ -9,5 +9,5 @@ cdef extern from *:
 new Foo(1, 2)
 
 _ERRORS = u"""
-9:7: Call with wrong number of arguments (expected 1, got 2)
+9:7: Call with wrong number of arguments (expected 0, got 2)
```
Comment (scoder) [inline tests/errors/cpp_no_constructor.pyx:12]: Yeah, this is what I meant with deterministic output above. This change seems arbitrary.

**Comment by scoder** [comment]
This was probably by accident, but please remember cleaning up the commit comment when merging. The one in https://github.com/cython/cython/commit/3d41922a23793fef61eba6b51b779d24bac9a1f1 contains a lot of unnecessary editing comments from this PR.

**Comment by vyasr** [comment]
@scoder @da-woods I ran into a couple of problems today that I was able to trace back to this PR. I was able to fix both of them, and I wasn't sure whether to open issues because I strongly suspect that they're simply unsupported use cases that happened to work previously and are better caught after the changes in this PR, so I thought I'd ask here first rather than opening an issue. Apologies if that's not your preferred approach, I wasn't sure what would be easier for you.

The first problem came from code that was declaring C++ template APIs in a pxd file as explicit template instantiations rather than as templates, i.e. roughly this
```
from libcpp.memory cimport unique_ptr
cdef extern from *:
    """
    #include <memory>

    template <typename T>
    class myclass {};

    template <typename T>
    void std::unique_ptr<myclass<T>> f() {}

    template class myclass<T>;
    template std::unique_ptr<myclass<T>>
    """"
    cdef cppclass myclass[T]:
        pass

    cdef unique_ptr[myclass[int]] f[int]() # Note the explicit `int` instead of `T` here
```

There's something more to it than just that since this MRE compiles for me, but changing the function declaration to use a template typename instead of a real type fixed compilation, and I've never seen the real type used there before, so I don't even know if this was intended to work.

The second problem came from a case where some code was shadowing a (non-templated) C function with a template C++ function:
```
from libc.stdlib cimport free

cdef extern from *:
    """
    template <typename T>
    void free(...) {}
    """
    cdef void free[T](...)

free[int](...)  # This used to work, but started throwing errors after this PR.
```

I wouldn't expect this to be valid Cython either, but apparently it worked before.

Do either of these changes appear to merit any follow-up to you? If so, I'm happy to open up an issue or two with proper MREs, but I wasn't sure if that was worthwhile here. Thanks!

**Comment by da-woods** [comment]
I'll have a look into them. It'll be a little while before this makes it into a release so we have time to look into this stuff.

**Comment by vyasr** [comment]
Sounds good. I will try to put together an MWE for these if possible and post them on the new issues you created.
